system/lib/wasmfs/streams.h

```diff
@@ -32,40 +32,56 @@ class StdinFile : public DataFile {
   static std::shared_ptr<StdinFile> getSingleton();
 };
 
-class StdoutFile : public DataFile {
+// A standard stream that writes: stdout or stderr.
+class WritingStdFile : public DataFile {
+protected:
   std::vector<char> writeBuffer;
 
-  __wasi_errno_t write(const uint8_t* buf, size_t len, off_t offset) override;
-
   __wasi_errno_t read(uint8_t* buf, size_t len, off_t offset) override {
     return __WASI_ERRNO_INVAL;
   };
 
-  // /dev/stdout reports a size of 0 in the terminal.
+  // /dev/stdout|stderr reports a size of 0 in the terminal.
   size_t getSize() override { return 0; }
 
 public:
-  StdoutFile(mode_t mode) : DataFile(mode, NullBackend, S_IFCHR) {}
-  static std::shared_ptr<StdoutFile> getSingleton();
+  WritingStdFile() : DataFile(S_IWUGO, NullBackend, S_IFCHR) {}
 };
 
-class StderrFile : public DataFile {
-  std::vector<char> writeBuffer;
-
-  // TODO: May not want to proxy stderr (fd == 2) to the main thread.
-  // This will not show in HTML - a console.warn in a worker is sufficient.
-  // This would be a change from the current FS.
+class StdoutFile : public WritingStdFile {
   __wasi_errno_t write(const uint8_t* buf, size_t len, off_t offset) override;
 
-  __wasi_errno_t read(uint8_t* buf, size_t len, off_t offset) override {
-    return __WASI_ERRNO_INVAL;
-  };
+public:
+  StdoutFile() {}
 
-  // /dev/stderr reports a size of 0 in the terminal.
-  size_t getSize() override { return 0; }
+  static std::shared_ptr<StdoutFile> getSingleton();
+
+  void flush() {
+    // Flush the musl libc buffers.
+    fflush(stdout);
+
+    // Write a null to flush the output.
+    const uint8_t nothing = 0;
+    write(&nothing, 1, 0);
+  }
+};
+
+class StderrFile : public WritingStdFile {
+  __wasi_errno_t write(const uint8_t* buf, size_t len, off_t offset) override;
 
 public:
-  StderrFile(mode_t mode) : DataFile(mode, NullBackend, S_IFCHR) {}
+  StderrFile() {}
+
   static std::shared_ptr<StderrFile> getSingleton();
+
+  void flush() {
+    // Flush the musl libc buffers.
+    fflush(stderr);
```

Comment:
Locally don't you want to be flushing *all* streams on exit.   e.g. a global call to `fflush(0)` somewhere?   It seems like this should ideally be handles in the some global place.  

Presumably this would not be needed at all if we integrated musl `exit()` code which calls `fflush(0)`?   Perhaps mention that since I hope we can do that soon.

Also, this is not conditional on EXIT_RUNTIME, but will run in all cases right?  Is that OK?

Reply:
Adding `fflush(0)`, sg, and a TODO for musl's exit.

Without `EXIT_RUNTIME` we do not call atexits, which means that the destructor of WasmFS is not called. So this just applies to `EXIT_RUNTIME`